r/R/dplyr-funcs-datetime.R

```diff
@@ -376,6 +376,21 @@ register_bindings_duration_helpers <- function() {
   register_binding("dyears", function(x = 1) {
     make_duration(x * 31557600, "s")
   })
+  register_binding("dseconds", function(x = 1) {
+    make_duration(x, "s")
+  })
+  register_binding("dmilliseconds", function(x = 1) {
+    make_duration(x, "ms")
+  })
+  register_binding("dmicroseconds", function(x = 1) {
+    make_duration(x, "us")
+  })
+  register_binding("dnanoseconds", function(x = 1) {
+    make_duration(x, "ns")
+  })
+  register_binding("dpicoseconds", function(x = 1) {
+    abort("Duration in picoseconds not supported in Arrow.")
+  })
```

Comment:
I think we should keep this as is written out like this, but this is right on the border where we might consider doing the mapping between function names and durations programmatically call `register_binding()`.

Reply:
I would like to try with programmatically mapping function names and `register_binding()`, to see how it goes =)

Reply:
I am a bit stuck, any pointers where to look for what's wrong in this try of the mapping?

```R
.helpers_function_map <- list(
  "dminutes" = c(60, "s"),
  "dhours" = c(3600, "s"),
  "ddays" = c(86400, "s"),
  "dweeks" = c(604800, "s"),
  "dmonths" = c(2629800, "s"),
  "ydyears" = c(31557600, "s"),
  "dseconds" = c(1, "s"),
  "dmilliseconds" = c(1, "ms"),
  "dmicroseconds" = c(1, "us"),
  "dnanoseconds" = c(1, "ns")
)
make_duration <- function(x, unit) {
  x <- build_expr("cast", x, options = cast_options(to_type = int64()))
  x$cast(duration(unit))
}
register_bindings_duration_helpers <- function() {
  duration_helpers_map_factory <- function(value, unit) {
    force(value)
    force(unit)
    function(x = 1) make_duration(x * value, unit)
  }

  for (name in names(.helpers_function_map)){
    register_binding(
      name,
      duration_helpers_map_factory(
        .helpers_function_map[[name]][1],
        .helpers_function_map[[name]][2]
      )
    )
  }

  register_binding("dpicoseconds", function(x = 1) {
    abort("Duration in picoseconds not supported in Arrow.")
  })
}
```

Reply:
When you say "where to look for what's wrong", how do you know something is wrong?

Reply:
The tests are failing and they have passed before :(

Reply:
Yeah, it would be good to post an output 🤦‍♀️ 

```R
> devtools::test(filter="datetime")
ℹ Loading arrow
Some features are not enabled in this build of Arrow. Run `arrow_info()` for more information.
ℹ Testing arrow
Some features are not enabled in this build of Arrow. Run `arrow_info()` for more information.
✔ | F W S  OK | Context
✖ | 7     381 | dplyr-funcs-datetime [2.0s]                                                                                                                  
─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
Failure (test-dplyr-funcs-datetime.R:1263:3): dminutes, dhours, ddays, dweeks, dmonths, dyears
`via_batch <- rlang::eval_tidy(expr, rlang::new_data_mask(rlang::env(.input = record_batch(tbl))))` threw an unexpected warning.
Message: Expression dyears(x) not supported in Arrow; pulling data into R
Class:   simpleWarning/warning/condition
Backtrace:
  1. arrow:::compare_dplyr_binding(...)
       at test-dplyr-funcs-datetime.R:1263:2
 12. arrow::mutate.ArrowTabular(...)
 13. arrow::abandon_ship(call, .data, msg)
       at r/R/dplyr-mutate.R:62:6

Failure (test-dplyr-funcs-datetime.R:1263:3): dminutes, dhours, ddays, dweeks, dmonths, dyears
`via_table <- rlang::eval_tidy(expr, rlang::new_data_mask(rlang::env(.input = arrow_table(tbl))))` threw an unexpected warning.
Message: Expression dyears(x) not supported in Arrow; pulling data into R
Class:   simpleWarning/warning/condition
Backtrace:
  1. arrow:::compare_dplyr_binding(...)
       at test-dplyr-funcs-datetime.R:1263:2
 12. arrow::mutate.ArrowTabular(...)
 13. arrow::abandon_ship(call, .data, msg)
       at r/R/dplyr-mutate.R:62:6

Failure (test-dplyr-funcs-datetime.R:1278:3): dminutes, dhours, ddays, dweeks, dmonths, dyears
`via_batch <- rlang::eval_tidy(expr, rlang::new_data_mask(rlang::env(.input = record_batch(tbl))))` threw an unexpected warning.
Message: Expression date_to_add + ddays - dhours(3) not supported in Arrow; pulling data into R
Class:   simpleWarning/warning/condition
Backtrace:
  1. arrow:::compare_dplyr_binding(...)
       at test-dplyr-funcs-datetime.R:1278:2
 12. arrow::mutate.ArrowTabular(...)
 13. arrow::abandon_ship(call, .data, msg)
       at r/R/dplyr-mutate.R:62:6

Failure (test-dplyr-funcs-datetime.R:1278:3): dminutes, dhours, ddays, dweeks, dmonths, dyears
`via_table <- rlang::eval_tidy(expr, rlang::new_data_mask(rlang::env(.input = arrow_table(tbl))))` threw an unexpected warning.
Message: Expression date_to_add + ddays - dhours(3) not supported in Arrow; pulling data into R
Class:   simpleWarning/warning/condition
Backtrace:
  1. arrow:::compare_dplyr_binding(...)
       at test-dplyr-funcs-datetime.R:1278:2
 12. arrow::mutate.ArrowTabular(...)
 13. arrow::abandon_ship(call, .data, msg)
       at r/R/dplyr-mutate.R:62:6

Failure (test-dplyr-funcs-datetime.R:1292:3): dminutes, dhours, ddays, dweeks, dmonths, dyears
`via_batch <- rlang::eval_tidy(expr, rlang::new_data_mask(rlang::env(.input = record_batch(tbl))))` threw an unexpected warning.
Message: Expression dminutes(1) not supported in Arrow; pulling data into R
Class:   simpleWarning/warning/condition
Backtrace:
  1. arrow:::compare_dplyr_binding(...)
       at test-dplyr-funcs-datetime.R:1292:2
 12. arrow::mutate.ArrowTabular(...)
 13. arrow::abandon_ship(call, .data, msg)
       at r/R/dplyr-mutate.R:62:6

Failure (test-dplyr-funcs-datetime.R:1292:3): dminutes, dhours, ddays, dweeks, dmonths, dyears
`via_table <- rlang::eval_tidy(expr, rlang::new_data_mask(rlang::env(.input = arrow_table(tbl))))` threw an unexpected warning.
Message: Expression dminutes(1) not supported in Arrow; pulling data into R
Class:   simpleWarning/warning/condition
Backtrace:
  1. arrow:::compare_dplyr_binding(...)
       at test-dplyr-funcs-datetime.R:1292:2
 12. arrow::mutate.ArrowTabular(...)
 13. arrow::abandon_ship(call, .data, msg)
       at r/R/dplyr-mutate.R:62:6

Error (test-dplyr-funcs-datetime.R:1312:3): dseconds, dmilliseconds, dmicroseconds, dnanoseconds, dpicoseconds
Error in `collect(.)`: NotImplemented: Function 'multiply_checked' has no kernel matching input types (array[int32], scalar[string])
/Users/alenkafrim/repos/arrow/cpp/src/arrow/compute/exec/expression.cc:340  call.function->DispatchBest(&descrs)
/Users/alenkafrim/repos/arrow/cpp/src/arrow/compute/exec/expression.cc:411  BindImpl(std::move(argument), in, shape, exec_context)
/Users/alenkafrim/repos/arrow/cpp/src/arrow/compute/exec/expression.cc:411  BindImpl(std::move(argument), in, shape, exec_context)
/Users/alenkafrim/repos/arrow/cpp/src/arrow/compute/exec/project_node.cc:67  expr.Bind(*inputs[0]->output_schema())
Backtrace:
  1. arrow:::compare_dplyr_binding(...)
       at test-dplyr-funcs-datetime.R:1312:2
 11. arrow::collect.arrow_dplyr_query(.)
─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────

══ Results ══════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════
Duration: 2.2 s

[ FAIL 7 | WARN 0 | SKIP 0 | PASS 381 ]
```

Reply:
Thanks @thisisnic for all the help! Will do a better job at asking questions next time (reprex and such).

The code needs a bit of tweaking and will be ready to commit.